File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int comp = 1;
        int incomp = 1;
        int index = -1;
        for(int i = 0; i < nums.size(); i++)
        {
            comp *= nums[i];
            if(nums[i] == 0 && index == -1)
                index = i;
            if(i != index)
                incomp *= nums[i];
        }
        vector<int> res(nums.size());
        for(int j = 0; j < nums.size(); j++)
        {
            if(nums[j] != 0)
                res[j] = comp/nums[j];
            else
                res[j] = incomp;
        }
        return res;
    }
};

File: Data Structures & Algorithms/daily-temperatures/submission-0.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        # add nums to stack and for each number check to see if new is larger
        # if new num is larger remove the top value from stack and account for the distance
        # for the stack it is a tuple with temp, and index

        res = [0] * len(temperatures)
        stack = []
        for i in range(len(temperatures)):
            temp = temperatures[i]
            
            while stack and stack[-1][0] < temp:
                index = stack[-1][1]
                stack.pop()
                res[index] = i - index 
            
            stack.append((temp, i))
        
        return res